### DISCUSSION TITLE
timing control during experimental procedures

### DISCUSSION THREAD

**Discussion by untzag** [pymeasure/pymeasure#564] [Q&A] [answered]
Disclaimer: I am totally new to pymeasure. Really cool project :smile:

From what I can tell, the accepted approach to timing control is to simply sleep for some amount of time between each experimental step, is that correct? Is there any built-in feature to know how long I need to wait before moving on? I'm especially thinking about motors.

**Comment by CasperSchippers** [comment, marked as answer]
For a lot of timing control situations I think indeed putting in a sleep is the standard way to go, but that can depend on the precise situation. 

From your question I guess you want to move a motor and want to wait until the motor is done moving (please correct me if I misinterpreted). How long, or how, you should wait I think depends on the precise motor controller; more simple motor controllers just require you to guestimate the amount of time you need to wait; with others you can query the movement speed and calculate from there; others even allow you to query whether the movement is finished, meaning that you just query this periodically until it says it is finished.

**Reply by untzag** [reply]
That makes sense---I see for example the `motion_done` property for the newport esp300. I guess I just need to figure it out for each hardware specifically.

Thanks!

**Reply by SBresler** [reply]
This does depend on how strict you need your timing to be.

I have thought about this and am wondering if it is possible to offload the clocking to a more precise device and set it up as an interrupt type event. This is for super critical timing applications.

**Reply by bilderbuchi** [reply]
For super critical timing you might indeed need something else - sleep(), depending on your OS, will only be precise on the order of 10ms, see https://stackoverflow.com/questions/1133857/how-accurate-is-pythons-time-sleep
